oscore-cf/src/main/java/org/eclipse/californium/oscore/ObjectSecurityLayer.java

```diff
@@ -0,0 +1,309 @@
+
+	/**
+	 * The list of non-protected resource names
+	 */
+	private List<String> unprotected;
+	
+	/**
+	 * Constructor
+	 * 
+	 * @param unprotected  the list of unprotected resources
+	 */
+	public ObjectSecurityLayer(List<String> unprotected) {
+		this.unprotected = unprotected;
+	}
+
+	/**
+	 * Encrypt an outgoing request using the OSCore context.
+	 * 
+	 * @param message  the message
+	 * @param ctx  the OSCore context
+	 * 
+	 * @return  the encrypted message
+	 * 
+	 * @throws OSException error while encrypting request
+	 */
+	public static Request prepareSend(Request message, OSCoreCtx ctx, int realCode) 
+			throws OSException {
+		return RequestEncryptor.encrypt(message, ctx, realCode);
+	}
+
+	/**
+	 * Encrypt an outgoing response using the OSCore context.
+	 * 
+	 * @param message  the message
+	 * @param ctx  the OSCore context
+	 * 
+	 * @return  the encrypted message
+	 * 
+	 * @throws OSException error while encrypting response
+	 */
+	public static Response prepareSend(Response message, OSCoreCtx ctx, int realCode, final boolean newPartialIV) 
+			throws OSException {
+		return ResponseEncryptor.encrypt(message, ctx, realCode, newPartialIV);
+	}
+
+	/**
+	 * Decrypt an incoming request using the right OSCore context
+	 * 
+	 * @param request  the incoming request
+	 * @param db  the OSCore context database
+	 * 
+	 * @return  the cid of the OSCore context used for this request
+	 * 
+	 * @throws OSException  error while decrypting request
+	 */
+	public static byte[] prepareReceive(Request request) 
+			throws OSException {
+		return RequestDecryptor.decrypt(request);
+	}
+
+	/**
+	 * Decrypt an incoming response using the right OSCore context
+	 * 
+	 * @param response  the incoming request
+	 * @param db  the OSCore context database
+	 * @return  the decrypted and verified response
+	 * 
+	 * @throws OSException error while decrypting response
+	 */
+	public static Response prepareReceive(Response response) 
+			throws OSException {
+		return ResponseDecryptor.decrypt(response);
+	}
+
+	@Override
+	public void sendRequest(Exchange exchange, Request request){
+		Request req = request;
+		if(shouldProtectRequest(request)){
+			try {
+				String uri = request.getURI();             
+				OSCoreCtxDB db = HashMapCtxDB.getInstance();
+				
+				if(uri == null){
+					LOGGER.severe(OSExceptionHandler.URI_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.URI_NULL);
+				}
+				if(db == null){
+					LOGGER.severe(OSExceptionHandler.DB_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.DB_NULL);
+				}
+				
+				OSCoreCtx ctx = db.getContext(uri);
+				if( ctx == null){
+					LOGGER.severe(OSExceptionHandler.CTX_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.CTX_NULL);
+				}
+
+				exchange.setCryptographicContextID(ctx.getRecipientId());
+				int seqByToken = ctx.getSenderSeq();
+				int realCode = request.getCode().value;				
```

Comment:
Why is the `realCode` extracted here and then passed to methods along with the request?

Reply:
Because the setFakeCodeRequest(request) creates a new Request with a different code. I need the realCode later.

Reply:
Up to the `RequestEncryptor`, the original request seems not to be changed.
Or did I oversee something. So up to that, the request will be enough. And there the real code may be processed.

Reply:
Uups, I overseen that `setFakeCodeRequest` is called before ...

I tried to check your data-flow.
Would it be possible, to create the new `Request` in `RequestEncryptor.encrypt`, leaving the original Request unchanged? Maybe that's simplifies it a little.

Reply:
And for the way back (or the response), I would also prefer to replace the codes and create new message, when encryption/decryption is done.

Reply:
It should not be a problem and I am starting to understand how it will improve readability.